src/ship_cmd.cpp

```diff
@@ -525,6 +525,44 @@ static const byte _ship_subcoord[4][6][3] = {
 	}
 };
 
+static int ShipTestUpDown(const Ship *v)
+{
+	if (!IsTileType(v->tile, MP_WATER) || !IsLock(v->tile)) return 0;
+
+	int x = v->x_pos & 0xF;
+	int y = v->y_pos & 0xF;
+	if (x != 8 && y != 8) return 0;
+
+	DiagDirection diagdir = GetInclinedSlopeDirection(GetTileSlope(v->tile));
+	if (diagdir == INVALID_DIAGDIR) return 0;
+
+	switch (diagdir) {
```

Comment:
I don't know if it feasible, but may this switch be changed to something like...
`assert(IsValidDiagdir(diagdir));
return (DirToDiagdir(v->direction) == diagdir) ? 1 : -1; `
As a ship can only move along the direction of the lock and I think rotation is dealt earlier.
Just asking here.

Reply:
Yes, thanks, it's a lot smaller.